File: scripts/generate-og-image.mjs
#!/usr/bin/env node
/**
 * Generate default Open Graph image for Cache McClure website.
 * Creates a 1200x630px image with site branding using Playwright.
 */

import { chromium } from 'playwright';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { writeFileSync, mkdirSync } from 'fs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const projectRoot = dirname(__dirname);

// HTML template for the OG image
const htmlTemplate = `
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>OG Image</title>
  <style>
    * {
      margin: 0;
      padding: 0;
      box-sizing: border-box;
    }

    body {
      width: 1200px;
      height: 630px;
      background: linear-gradient(135deg, #0f172a 0%, #1e293b 100%);
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
      font-family: 'Arial', 'Helvetica', sans-serif;
      position: relative;
      overflow: hidden;
    }

    /* Decorative corner elements */
    .corner {
      position: absolute;
      width: 60px;
      height: 60px;
      border: 4px solid #6366f1;
    }

    .corner.top-left {
      top: 80px;
      left: 80px;
      border-right: none;
      border-bottom: none;
    }

    .corner.top-right {
      top: 80px;
      right: 80px;
      border-left: none;
      border-bottom: none;
    }

    .corner.bottom-left {
      bottom: 80px;
      left: 80px;
      border-right: none;
      border-top: none;
    }

    .corner.bottom-right {
      bottom: 80px;
      right: 80px;
      border-left: none;
      border-top: none;
    }

    /* Horizontal accent lines */
    .accent-line {
      position: absolute;
      left: 100px;
      right: 100px;
      height: 4px;
      background: linear-gradient(90deg, transparent, #6366f1, transparent);
    }

    .accent-line.top {
      top: 195px;
    }

    .accent-line.bottom {
      bottom: 195px;
    }

    /* Content container */
    .content {
      text-align: center;
      z-index: 10;
    }

    .title {
      font-size: 90px;
      font-weight: bold;
      color: #f1f5f9;
      letter-spacing: 8px;
      margin-bottom: 20px;
      text-shadow: 0 0 30px rgba(99, 102, 241, 0.5);
    }

    .subtitle {
      font-size: 42px;
      color: #94a3b8;
      letter-spacing: 3px;
      font-weight: 300;
    }

    /* Sci-fi grid background */
    .grid {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      background-image:
        linear-gradient(rgba(99, 102, 241, 0.03) 1px, transparent 1px),
        linear-gradient(90deg, rgba(99, 102, 241, 0.03) 1px, transparent 1px);
      background-size: 50px 50px;
      opacity: 0.3;
    }

    /* Glow effect */
    .glow {
      position: absolute;
      width: 400px;
      height: 400px;
      border-radius: 50%;
      background: radial-gradient(circle, rgba(99, 102, 241, 0.15) 0%, transparent 70%);
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      z-index: 1;
    }
  </style>
</head>
<body>
  <div class="grid"></div>
  <div class="glow"></div>

  <div class="corner top-left"></div>
  <div class="corner top-right"></div>
  <div class="corner bottom-left"></div>
  <div class="corner bottom-right"></div>

  <div class="accent-line top"></div>
  <div class="accent-line bottom"></div>

  <div class="content">
    <div class="title">CACHE McCLURE</div>
    <div class="subtitle">Science Fiction Author</div>
  </div>
</body>
</html>
`;

async function generateOGImage() {
  console.log('🚀 Launching browser...');
  const browser = await chromium.launch();
  const page = await browser.newPage();

  // Set viewport to OG image dimensions
  await page.setViewportSize({ width: 1200, height: 630 });

  console.log('📄 Loading HTML template...');
  await page.setContent(htmlTemplate);

  // Wait for fonts and styles to load
  await page.waitForTimeout(1000);

  console.log('📸 Capturing screenshot...');
  const screenshot = await page.screenshot({
    type: 'jpeg',
    quality: 90,
  });

  await browser.close();

  // Save the image
  const outputPath = join(projectRoot, 'public', 'cache-mcclure-og.jpg');
  writeFileSync(outputPath, screenshot);

  const fileSizeKB = (screenshot.length / 1024).toFixed(1);

  console.log(`✅ Created Open Graph image: ${outputPath}`);
  console.log(`   Dimensions: 1200x630px`);
  console.log(`   File size: ${fileSizeKB}KB`);

  if (parseFloat(fileSizeKB) > 300) {
    console.log(`   ⚠️  Warning: File size exceeds recommended 300KB`);
  } else {
    console.log(`   ✅ File size is within recommended limits`);
  }
}

// Run the generator
generateOGImage().catch(error => {
  console.error('❌ Error generating OG image:', error);
  process.exit(1);
});

File: scripts/generate-og-image.py
#!/usr/bin/env python3
"""
Generate default Open Graph image for Cache McClure website.
Creates a 1200x630px image with site branding.
"""

from PIL import Image, ImageDraw, ImageFont
import os

# Image dimensions (Open Graph standard)
WIDTH = 1200
HEIGHT = 630

# Color scheme (matching the website's sci-fi aesthetic)
BACKGROUND_COLOR = (15, 23, 42)  # Dark slate (tailwind slate-900)
PRIMARY_COLOR = (148, 163, 184)  # Light slate (tailwind slate-400)
ACCENT_COLOR = (99, 102, 241)  # Indigo (tailwind indigo-500)
TEXT_COLOR = (241, 245, 249)  # Very light slate (tailwind slate-100)

def create_og_image():
    """Create the Open Graph image."""
    # Create image with dark background
    img = Image.new('RGB', (WIDTH, HEIGHT), BACKGROUND_COLOR)
    draw = ImageDraw.Draw(img)

    # Try to use system fonts, fall back to default
    try:
        # Try to find a modern sans-serif font
        font_paths = [
            '/System/Library/Fonts/Supplemental/Arial Bold.ttf',  # macOS
            '/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf',  # Linux
            'C:\\Windows\\Fonts\\arialbd.ttf',  # Windows
        ]

        title_font = None
        subtitle_font = None

        for font_path in font_paths:
            if os.path.exists(font_path):
                title_font = ImageFont.truetype(font_path, 80)
                subtitle_font = ImageFont.truetype(font_path, 40)
                break

        if not title_font:
            # Fallback to default font
            title_font = ImageFont.load_default()
            subtitle_font = ImageFont.load_default()

    except Exception:
        # If all else fails, use default
        title_font = ImageFont.load_default()
        subtitle_font = ImageFont.load_default()

    # Add decorative elements (sci-fi aesthetic)
    # Draw accent lines
    line_y = HEIGHT // 2 - 120
    draw.rectangle([100, line_y, 1100, line_y + 4], fill=ACCENT_COLOR)

    line_y = HEIGHT // 2 + 120
    draw.rectangle([100, line_y, 1100, line_y + 4], fill=ACCENT_COLOR)

    # Add corner accents
    corner_size = 40
    corner_thickness = 4

    # Top-left corner
    draw.rectangle([80, 80, 80 + corner_size, 80 + corner_thickness], fill=ACCENT_COLOR)
    draw.rectangle([80, 80, 80 + corner_thickness, 80 + corner_size], fill=ACCENT_COLOR)

    # Top-right corner
    draw.rectangle([WIDTH - 80 - corner_size, 80, WIDTH - 80, 80 + corner_thickness], fill=ACCENT_COLOR)
    draw.rectangle([WIDTH - 80 - corner_thickness, 80, WIDTH - 80, 80 + corner_size], fill=ACCENT_COLOR)

    # Bottom-left corner
    draw.rectangle([80, HEIGHT - 80 - corner_thickness, 80 + corner_size, HEIGHT - 80], fill=ACCENT_COLOR)
    draw.rectangle([80, HEIGHT - 80 - corner_size, 80 + corner_thickness, HEIGHT - 80], fill=ACCENT_COLOR)

    # Bottom-right corner
    draw.rectangle([WIDTH - 80 - corner_size, HEIGHT - 80 - corner_thickness, WIDTH - 80, HEIGHT - 80], fill=ACCENT_COLOR)
    draw.rectangle([WIDTH - 80 - corner_thickness, HEIGHT - 80 - corner_size, WIDTH - 80, HEIGHT - 80], fill=ACCENT_COLOR)

    # Add text
    title = "CACHE McCLURE"
    subtitle = "Science Fiction Author"

    # Calculate text positions (centered)
    title_bbox = draw.textbbox((0, 0), title, font=title_font)
    title_width = title_bbox[2] - title_bbox[0]
    title_x = (WIDTH - title_width) // 2
    title_y = HEIGHT // 2 - 50

    subtitle_bbox = draw.textbbox((0, 0), subtitle, font=subtitle_font)
    subtitle_width = subtitle_bbox[2] - subtitle_bbox[0]
    subtitle_x = (WIDTH - subtitle_width) // 2
    subtitle_y = HEIGHT // 2 + 20

    # Draw text
    draw.text((title_x, title_y), title, font=title_font, fill=TEXT_COLOR)
    draw.text((subtitle_x, subtitle_y), subtitle, font=subtitle_font, fill=PRIMARY_COLOR)

    # Save image
    output_path = os.path.join(
        os.path.dirname(os.path.dirname(__file__)),
        'public',
        'cache-mcclure-og.jpg'
    )

    # Save as high-quality JPEG
    img.save(output_path, 'JPEG', quality=90, optimize=True)

    print(f"✓ Created Open Graph image: {output_path}")
    print(f"  Dimensions: {WIDTH}x{HEIGHT}px")

    # Check file size
    file_size = os.path.getsize(output_path)
    file_size_kb = file_size / 1024
    print(f"  File size: {file_size_kb:.1f}KB")

    if file_size_kb > 300:
        print(f"  ⚠ Warning: File size exceeds recommended 300KB")
    else:
        print(f"  ✓ File size is within recommended limits")

if __name__ == '__main__':
    try:
        create_og_image()
    except ImportError:
        print("Error: PIL/Pillow is not installed.")
        print("Install it with: pip3 install Pillow")
        exit(1)
    except Exception as e:
        print(f"Error creating Open Graph image: {e}")
        exit(1)
